### PR TITLE
fix numerical issue in function score query

### PR DESCRIPTION
**Pull request by brwe**
we should sum the weights as double to not lose precision. also,
the tests should simulate exactly what function score does and then test
for equality of scores.

Also, while at it I changed the same thing for explain. 
FiltersFunctionScoreQuery sums up scores and weights and scores as double but when
we explain we cannot get the double scores from the explanation of score
functions. as a result we cannot compute the exact score from the explanations
of the functions alone.
this commit makes the explanation more accurate but also causes the score to be
computed one additional time.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/lucene/search/function/FiltersFunctionScoreQuery.java [hunk 1]
```diff
@@ -232,44 +224,11 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
...
-                }
-                if (weightSum != 0) {
-                    factor = totalFactor;
-                    if (scoreMode == ScoreMode.AVG) {
-                        factor /= weightSum;
-                    }
-                }
-            }
+            FiltersFunctionFactorScorer scorer = (FiltersFunctionFactorScorer)scorer(context);
+            scorer.advance(doc);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/lucene/search/function/FiltersFunctionScoreQuery.java]: should we at least assert that scorer.advance(doc) returned doc?

**Comment by jpountz** [comment]
> also causes the score to be computed one additional time.

This is only the case when calling explain(), right? If so this is no issue at all.

**Comment by brwe** [comment]
> This is only the case when calling explain(), right? 

yes, everything else stays the same

I pushed a new commit.

**Comment by jpountz** [comment]
LGTM

**Comment by brwe** [comment]
@jpountz thanks for the quick review!

**Comment by nik9000** [comment]
Thanks @brwe ! Much better!

**Pull request merged**